src/browser/ChromeListener.cpp

```diff
@@ -0,0 +1,614 @@
+            QString login = decrypted.value("login").toString();
+            QString password = decrypted.value("password").toString();
+            QString submitUrl = decrypted.value("submitUrl").toString();
+            QString uuid = decrypted.value("uuid").toString();
+            QString realm = ""; // ?
+            QMutexLocker locker(&m_mutex);
+            if (uuid.isEmpty())
+                m_service.addEntry(id, login, password, url, submitUrl, realm);
+            else
+                m_service.updateEntry(id, uuid, login, password, url);
+
+            QJsonObject message;
+            message["count"] = QJsonValue::Null;
+            message["entries"] = QJsonValue::Null;
+            message["error"] = "";
+            message["hash"] = hash;
+            message["version"] = KEEPASSX_VERSION;
+            message["success"] = "true";
+            message["nonce"] = nonce;
+
+            QString replyMessage(QJsonDocument(message).toJson());
+            QJsonObject response;
+            response["action"] = action;
+            response["message"] = encrypt(replyMessage, nonce);
+            response["nonce"] = nonce;
+
+            sendReply(response);
+        }
+    }
+}
+
+void ChromeListener::sendReply(const QJsonObject json)
+{
+    QString reply(QJsonDocument(json).toJson());
+    uint len = reply.length();
+
+    std::cout << char(((len>>0) & 0xFF))
+                << char(((len>>8) & 0xFF))
+                << char(((len>>16) & 0xFF))
+                << char(((len>>24) & 0xFF));
+    std::cout << reply.toStdString() << std::flush;
+
+    if (BrowserSettings::supportBrowserProxy()) {
+        m_udpSocket.writeDatagram(reply.toUtf8(), m_peerAddr, m_peerPort);
+    }
+}
+
+void ChromeListener::sendErrorReply(const QString &action, const int errorCode)
+{
+    QJsonObject response;
+    response["action"] = action;
+    response["errorCode"] = QString::number(errorCode);
+    response["error"] = getErrorMessage(errorCode);
+    sendReply(response);
+}
+
+QString ChromeListener::getErrorMessage(const int errorCode) const
+{
+    switch(errorCode) {
+        case ERROR_KEEPASS_DATABASE_NOT_OPENED:             return "Database not opened";
+        case ERROR_KEEPASS_DATABASE_HASH_NOT_RECEIVED:      return "Database hash not available";
+        case ERROR_KEEPASS_CLIENT_PUBLIC_KEY_NOT_RECEIVED:  return "Client public key not received";
+        case ERROR_KEEPASS_CANNOT_DECRYPT_MESSAGE:          return "Cannot decrypt message";
+        case ERROR_KEEPASS_TIMEOUT_OR_NOT_CONNECTED:        return "Timeout or cannot connect to KeePassXC";
+        case ERROR_KEEPASS_ACTION_CANCELLED_OR_DENIED:      return "Action cancelled or denied";
+        default:                                            return "Unknown error";
+    }
+}
+
+QString ChromeListener::encrypt(const QString decrypted, const QString nonce) const
+{
+    QString result;
+    unsigned char n[crypto_box_NONCEBYTES];
```

Comment:
You can avoid these duplicate arrays altogether if you just use QByteArray::data() in line 517.